internal/controllers/machine/machine_controller_status.go

```diff
@@ -512,6 +515,101 @@ func (c machineConditionCustomMergeStrategy) Merge(conditions []v1beta2condition
 	).Merge(conditions, conditionTypes)
 }
 
+// transformControlPlaneAndEtcdConditions Group readiness gates for control plane conditions and etcd when they have the same messages.
+// Note: the implementation is based on KCP conditions, but ideally other control plane implementation could
+// take benefit from this optimization by naming conditions with APIServer, ControllerManager, Scheduler or the Etcd prefix.
+func transformControlPlaneAndEtcdConditions(messages []string) []string {
+	isControlPlaneCondition := func(c string) bool {
+		if strings.HasPrefix(c, "* APIServer") {
+			return true
+		}
+		if strings.HasPrefix(c, "* ControllerManager") {
+			return true
+		}
+		if strings.HasPrefix(c, "* Scheduler") {
+			return true
+		}
+		// Note. Etcd pod healthy is considered as part of control plane components in KCP
+		// because it is not part of the etcd cluster.
+		// Might be in future we want to make this check more strictly tight to KCP machines e.g. by checking the machine's owner;
+		// for now, we consider checking for the exact condition name an acceptable trade off (same below).
+		if c == "* EtcdPodHealthy:" {
+			return true
+		}
+		return false
+	}
+
+	out1 := []string{}
+	controlPlaneConditionsCount := 0
+	controlPlaneMsg := ""
+	for _, m := range messages {
+		sep := strings.Index(m, ":")
+		if sep == -1 {
+			out1 = append(out1, m)
+			continue
+		}
+
+		c, msg := m[:sep], m[sep+1:]
+		if !isControlPlaneCondition(c) {
```

Comment:
This does not work for `EtcdPodHealthy` if my thinking is right (`:` is missing). Same below

Reply:
Note: propably good to have test coverage for EtcdPodHealthy

Reply:
Good catch, fixed + added test coverage